### DISCUSSION TITLE
Base64 Content Transfer Encoding

### DISCUSSION THREAD

**Discussion by MakakWasTaken** [phax/phase4#163] [General]
I have recently started receiving the error: Error Reading Input Stream.

I suspect that it might be caused by the partner switching from sending the attachment using the binary Content-Transfer-Encoding to base64.

Is it possible to receive documents using this encoding?

**Comment by phax** [comment]
Was the `Content-Transfer-Encoding` set on a specific MIME part, of was it set on the full message?

**Comment by MakakWasTaken** [comment]
The raw content of the transaction looks like the following (With soap envelope and attachment redacted):

```
--=-VahA61BNdImrmvkKe6eQ4A==
Content-Type: application/soap+xml; charset=utf-8

** SOAP ENVELOPE **
--=-VahA61BNdImrmvkKe6eQ4A==
Content-Type: application/octet-stream
Content-Id: <58d2a43d-3f24-4544-9f30-766f78988b12.edi>
Content-Disposition: attachment
Content-Transfer-Encoding: base64

**BASE64 STRING**

--=-VahA61BNdImrmvkKe6eQ4A==--
``` 

**EDIT:**

OBS. I am not sure if this specifically is causing the overall problem that I am having, so I'll provide some more information:
The below error started happening after our partner switched to a new system. The only difference I noticed in the multipart was that the `Content-Transfer-Encoding` had switched from `binary` to `base64`. So I concluded that this was the cause of the error.

The exception is:
`com.helger.web.multipart.MultipartMalformedStreamException: Stream ended unexpectedly`

I am not sure why this error suddenly started happening. It is the same files that they used to send before the update.

<details>
<summary>Full Stacktrace</summary>
<br>

```
Internal error processing AS4 request
Technical details:
javax.mail.MessagingException: Error reading input stream;
  nested exception is:
	com.helger.web.multipart.MultipartMalformedStreamException: Stream ended unexpectedly
1.: javax.mail.internet.MimeBodyPart.<init>(MimeBodyPart.java:172)
2.: com.helger.phase4.servlet.AS4IncomingHandler.parseAS4Message(AS4IncomingHandler.java:215)
3.: com.helger.phase4.servlet.AS4RequestHandler.handleRequest(AS4RequestHandler.java:1725)
4.: com.helger.phase4.servlet.AS4RequestHandler.handleRequest(AS4RequestHandler.java:1764)
5.: com.helger.phase4.servlet.AS4XServletHandler.handleRequest(AS4XServletHandler.java:309)
6.: com.helger.phase4.servlet.AS4XServletHandler.handleRequest(AS4XServletHandler.java:343)
7.: com.helger.xservlet.handler.simple.XServletHandlerToSimpleHandler.onRequest(XServletHandlerToSimpleHandler.java:245)
8.: com.helger.xservlet.AbstractXServlet._invokeHandler(AbstractXServlet.java:345)
9.: com.helger.xservlet.AbstractXServlet.service(AbstractXServlet.java:533)
10.: javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
11.: com.helger.xservlet.AbstractXServlet.service(AbstractXServlet.java:587)
  [34 elements omitted -- org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)]
==> [1] caused by com.helger.web.multipart.MultipartMalformedStreamException: Stream ended unexpectedly
1.: com.helger.web.multipart.MultipartStream$MultipartItemInputStream._makeAvailable(MultipartStream.java:820)
2.: com.helger.web.multipart.MultipartStream$MultipartItemInputStream.read(MultipartStream.java:705)
3.: java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:252)
4.: java.base/java.io.BufferedInputStream.read1(BufferedInputStream.java:292)
5.: java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:351)
6.: com.sun.mail.util.ASCIIUtility.getBytes(ASCIIUtility.java:278)
7.: javax.mail.internet.MimeBodyPart.<init>(MimeBodyPart.java:170)
  [44 elements omitted -- com.helger.phase4.servlet.AS4IncomingHandler.parseAS4Message(AS4IncomingHandler.java:215)]
```

</details>

**Comment by MakakWasTaken** [comment]
After some more research I have discovered that the partner also switched from sending with `CRLF` to `LF` line endings. According to this [post](https://stackoverflow.com/questions/53500627/malformedstreamexception-stream-ended-unexpectedly) this could also be the cause of the problem. 

To this extent I have some questions:
1. Is it possible to receive base64 encoded attachments already?
2. Is it possible to set up receiving for both types of line endings?

**Reply by phax** [reply]
Hi @MakakWasTaken  I haven't dived into this issue, as I am currently struggeling with other issues. I think in general, the `Content-Transfer-Encoding` is currently not honoured, and only `binary` is supported - but I haven't checked in the code - sorry. Please stay tuned....

**Reply by phax** [reply]
Are you getting a specific exception that you can share here or send to me via email?

**Reply by MakakWasTaken** [reply]
I have mentioned the exception that I am getting in the edited reply above.

**Reply by phax** [reply]
Sorry. Thanks. Mea culpa. Yes, this is a stripped down version of the Apache commons-fileupload.... checking

**Reply by MakakWasTaken** [reply]
As I understand it, the problem lies within [MultipartStream](https://github.com/phax/ph-web/blob/master/ph-web/src/main/java/com/helger/web/multipart/MultipartStream.java). As I understand it the class is only able to handle CRLF line endings. 

Some possibilities could be:
- Either allow editing `MultipartStream.HEADER_SEPARATOR`, `FIELD_SEPERATOR` & `BOUNDARY_PREFIX`. This could be done using a setLineEndings function, that allow either 'CRLF' or 'LF' on a per document basis. This would however also require editing `AS4IncomingHandler.java` to determine the line ending in the `aRequestIS`
- Another alternative could be to enforce CRLF in the files. This could be done using regex:
  ```java
  str.replaceAll("\r?\n", "\r\n")
  ```
  This would however require converting the `aRequestIS` to a `String` and then back into an `InputStream`. These solutions seem hacky, and it might be a better approach to just support using LF in MultipartStream. 
I am not sure if it is possible, but it could maybe be done by setting the following:
```java

  private static final byte [] HEADER_SEPARATOR = { LF, LF };
  private static final byte [] FIELD_SEPARATOR = { LF };
  private static final byte [] BOUNDARY_PREFIX = { LF, DASH, DASH };
```
And then just ignore the bytes that are equal to `CR` when reading from the inputstream?

**Reply by phax** [reply]
Hi Markus,
I think it is a bad idea to adopt the parser. In this particular case, the sender seems to be the one that is not compliant.
If you look at [RFC 2045](https://datatracker.ietf.org/doc/html/rfc2045) it is clearly defined HOW such a message should be formatted.
By introducing these kinds of "workarounds" into the parser it would in the end lower the interoperability and nobody would gain something, except complexity and confusion :)

What software is yout sender side using? Can that be deduced from the message ID or so? I really strongly suggest to contact their support to fix the message layout - they will run into similar issues with other AS4 solutions as well....

**Reply by MakakWasTaken** [reply]
That makes sense, I wasn't aware that there was a standard for the line endings 😄. I'll forward the standard to them and ask if they're able to comply. 

I think they have developed their own solution, but I am not 100% sure. I have a name of the system, but I am not sure i'm allowed to post it here. I can send it in an email if you'd like it.

**Reply by phax** [reply]
Yes, please send it to me via email. Feel free to also put me in CC when escalating issues - I am happy to help on the technical issues...

**Reply by MakakWasTaken** [reply]
I got my partner convinced to change their system using the above mentioned standard 😄.

In regards to the original question on whether there is support for base64 encoded messages, I'll have a look when it is implemented (Should be implemented during this week or at the latest next week).

I'll keep you in mind if any more technical problems arise 😉 

Thanks again

**Reply by phax** [reply]
@MakakWasTaken  were you able to resolve the issue?

**Reply by MakakWasTaken** [reply]
Yes, it seems to be working fine with the base64 encoding on the attachment.

The other issue has also been resolved.

**Comment by arj03** [comment]
Ran into the same kind of error, content must be CRLF encoded and not LF
